torchao/quantization/quantize_/workflows/int4/int4_marlin_sparse_tensor.py

```diff
@@ -0,0 +1,231 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD 3-Clause license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+from typing import List
+
+import torch
+
+from torchao.quantization.quant_primitives import (
+    MappingType,
+    choose_qparams_affine,
+    quantize_affine,
+)
+from torchao.utils import TorchAOBaseTensor
+
+__all__ = [
+    "Int4MarlinSparseTensor",
+]
+
+aten = torch.ops.aten
+
+
+class Int4MarlinSparseTensor(TorchAOBaseTensor):
+    tensor_data_names = ["qdata", "scale", "zero_point", "meta"]
+    tensor_attribute_names = ["block_size", "num_bits", "shape"]
+
+    def __new__(cls, qdata, scale, zero_point, meta, block_size, num_bits, shape):
+        kwargs = {}
+        kwargs["device"] = qdata.device
+        kwargs["dtype"] = scale.dtype
+        kwargs["requires_grad"] = False
+        return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(self, qdata, scale, zero_point, meta, block_size, num_bits, shape):
+        self.qdata = qdata
+        self.scale = scale
+        self.zero_point = zero_point
+        self.meta = meta
+        self.block_size = block_size
+        self.num_bits = num_bits
+
+    def _quantization_type(self):
+        return f"shape={self.shape}, block_size={self.block_size}, device={self.device}"
+
+    @staticmethod
+    def pre_process(input: torch.Tensor) -> torch.Tensor:
```

Comment:
nit: we can remove this function and inline the code in from_hp directly I think, can keep the comments